### ISSUE TITLE
2fa prompt although --otp used

### ISSUE THREAD

**Issue by BlueCase** [adrienverge/openfortivpn#170]
```
openfortivpn -vvv --otp=111111
DEBUG:  Loaded config file "/etc/openfortivpn/config".
VPN account password: 
DEBUG:  Config host = "vpn.abc.de"
DEBUG:  Config realm = ""
DEBUG:  Config port = "443"
DEBUG:  Config username = "lalalala"
DEBUG:  Config password = "********"
DEBUG:  One-time password = "111111"
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
INFO:   Connected to gateway.
2factor authentication token: 
```

openfortivpn version: 1.5.0 (same with 1.4.0)
Token: FortiToken 200 HW OTP

Have someone an idea whats could be the error?
Any ideas how to debug this error?

**Comment by BlueCase** [comment]
Hi,
has nobody an idea how to solve this problem?

**Comment by skuti-is** [comment]
Hi @BlueCase I fail to see the error.
I see info and debug but no error.

Can you describe the problem and mabe share logs from the fortigate box.

**Comment by BlueCase** [comment]
@skuti-is error is, that openfortivpn still want a 2fa key although I already provided it via --otp
I will check if I can get the logs form your forti box

**Comment by mrbaseman** [comment]
There are two different mechanisms: one time passwords (which you can supply with the `--otp` command line parameter and which is processed in [try_otp_auth](https://github.com/adrienverge/openfortivpn/blob/master/src/http.c#L336) and two factor authentication which is processed in [auth_log_in](https://github.com/adrienverge/openfortivpn/blob/master/src/http.c#L537). Apart from that there are variants in which the password you are supposed to enter is glued together out of a fixed part and a token and neither of the two mechanisms above are used. Which one is used depends on the configuration of your vpn and of the accounts used for it.
To me it looks as if your box requests 2factor while you supply otp. The next question would be if we shall implement a mechanism to supply 2fa on the command line or if we should re-use the --otp parameter and try that value for 2fa as well (even though the mechanism behind the curtains is slightly different).

**Comment by BlueCase** [comment]
@mrbaseman 
Thanks for your help.
Based on your infos I changed the code in http.c auth_log_in as follow:
```             
 struct vpn_config *cfg = tunnel->config;        
                if (cfg->otp[0] == '\0') {
                        read_password("Two-factor authentication token: ", tokenresponse, 255);
                } else {
                        url_encode(tokenresponse, cfg->otp);    
                }
```

With that the authentication with --otp and also the nm plugin working.
But still the question if a separate command line switch is needed or not.

**Comment by BlueCase** [comment]
@mrbaseman 
Would ask if it's possible to get a permanent fix in one of the next releases?

**Comment by mrbaseman** [comment]
Yes, it makes sense to address this. @adrienverge, @DimitriPapadopoulos what do you think, should we add another command line option for the 2fa token, or shall we re-use --otp here?

**Comment by DimitriPapadopoulos** [comment]
Yes, a new option is probably cleaner.

Perhaps this shows we need better documentation. It's a pity there's not much public information on FortiGate device configuration (for example HTML documentation of the most recent release) we could refer to when explaining client-side configuration - would be much easier to understand, maintain and document...

**Comment by DimitriPapadopoulos** [comment]
There's a new pull requet #255 by @ilkundrak which seems to be addressing this issue. It looks a single option for both OTP and 2FA is easier in practice:
> The distinction between a mechanism OTP is requested from how 2FA token is requested
> doesn't provide much clue into what's the distinction. Assume they're the same thing, not used together.

**Comment by lkundrak** [comment]
Ah, I didn't notice this ticket. Is my understanding, that you've come to the same conclusion correct?
